### DISCUSSION TITLE
How to remove expiry-address for addresses matching "#"

### DISCUSSION THREAD

**Discussion by eanbiso** [arkmq-org/arkmq-org-broker-operator#853] [Q&A] [answered]
Hi support,
I would need to understand how I can change the "expiry-address" default config for addresses matching "#" (i.e. all the addresses, if no other more specific setting is configured), so that no expiryAddress is configured and expired messages are simply removed from the queue and dropped.
Should I specify an expiryAddress value equal to "" (empty string) in the Address Settings Artemis CRD definition?
I saw that, by default, if I do not specify it (see the screenshot below, where I only changed the expiry delay configurations on the addressSetting matching "#"), the broker.xml file still contains the "ExpiryQueue" value.

![image](https://github.com/artemiscloud/activemq-artemis-operator/assets/48474129/68758d7f-94d2-453f-ac9f-cca07b01cb05)

<img width="459" alt="image" src="https://github.com/artemiscloud/activemq-artemis-operator/assets/48474129/911d648b-6c0a-4042-8184-b0ea80a17375">

I wanted to try configuring the expiryAddress on the same addressSetting equal to "" ... am I on the right path?

Many thanks

**Comment by brusdev** [comment]
Hi @eanbiso, Apache ActiveMQ Artemis drops expired messages only when the expiry address value is `null` and there is no way set a broker setting to null by using the CR. A workaround could be to create a [custom init container](https://github.com/artemiscloud/activemq-artemis-operator/blob/main/docs/tutorials/initcontainer.md) but that feature could be deprecated in future so I would suggest you to open a RFE for Apache ActiveMQ Artemis to drop expired messages also when the expiry address value is an empty string.

**Reply by eanbiso** [reply]
Hi @brusdev, thanks for your quick response.
I have opened a RFE via this new discussion [thread](https://lists.apache.org/thread/dw1g5q4x8n720okgf6ph5c0rp6kfjon7).

**Reply by jbertram** [reply, marked as answer]
FWIW, this should already be possible. For example, this can be set in `broker.xml`:
```xml
<expiry-address/>
```
And that will "unset" the expiry-address for all matching addresses so that messages which expire from queues which match that address are dropped.

**Reply by brusdev** [reply]
@jbertram is right, also an empty `expiry-address` value works. I was checking the [wrong expire method](https://github.com/apache/activemq-artemis/blob/2.33.0/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java#L2131) that only checks for a null value while the [right expire method](https://github.com/apache/activemq-artemis/blob/2.33.0/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java#L3832) checks for a null or empty value.

**Comment by eanbiso** [comment]
Perfect, thanks.
In this case it should be possible also in the k8s context, setting the corresponding value to "" (empty string).
I will try that.
Many thanks,

Andrea
________________________________
Da: Justin Bertram ***@***.***>
Inviato: martedì 9 aprile 2024 20:48
A: artemiscloud/activemq-artemis-operator ***@***.***>
Cc: Andrea Bisogno ***@***.***>; Mention ***@***.***>
Oggetto: Re: [artemiscloud/activemq-artemis-operator] How to remove expiry-address for addresses matching "#" (Discussion #853)


FWIW, this should already be possible. For example, this can be set in broker.xml:

<expiry-address/>

And that will "unset" the expiry-address for all matching addresses so that messages which expire from queues which match that address are dropped.

**Reply by eanbiso** [reply]
I've tested it now (setting the corresponding value on the Artemis CRD to ""), and it works.
The resulting broker.xml contains:
```
<address-setting match="#">
   <expiry-address></expiry-address>
```

Many thanks

Andrea
